File: pi/srcs/1_domain/System.cpp
#include "System.hpp"

System::System() : _current_state(INITIALIZATION)
{
}

state System::get_state() const
{
	return _current_state;
}

Commands System::getCommands() const
{
	return _commands;
}

std::string System::getStateString() const
{
	if (_current_state == INITIALIZATION) return "INITIALIZATION";
	if (_current_state == READY)          return "READY";
	if (_current_state == PROCESSING)     return "PROCESSING";
	return "SENDING";
}

void System::setCommandToExecute(const std::string &command)
{
	_command_to_execute = command;
}

std::string System::getCommandToExecute() const
{
	return _command_to_execute;
}

void System::ready()
{
	_current_state = READY;
}

void System::processing()
{
	_current_state = PROCESSING;
}

void System::sending()
{
	_current_state = SENDING;
}

void System::reset()
{
	_current_state = INITIALIZATION;
}

File: computer/srcs/_1_domain/system.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class State(Enum):
    INITIALIZATION = "INITIALIZATION"
    READY          = "READY"
    PROCESSING     = "PROCESSING"
    SENDING        = "SENDING"


@dataclass(frozen=True)
class Commands:
    valid: tuple[str, ...] = field(default=(
        "PING",
        "PONG",
        "GET_STATUS",
        "GET_PICO_STATUS",
        "START_CAPTURE",
    ))


class System:
    def __init__(self) -> None:
        self._current_state:      State    = State.INITIALIZATION
        self._commands:           Commands = Commands()
        self._command_to_execute: str      = ""

    def get_state(self) -> State:
        return self._current_state

    def get_state_string(self) -> str:
        return self._current_state.value

    def get_commands(self) -> Commands:
        return self._commands

    def set_command_to_execute(self, command: str) -> None:
        self._command_to_execute = command

    def get_command_to_execute(self) -> str:
        return self._command_to_execute

    def ready(self) -> None:
        self._current_state = State.READY

    def processing(self) -> None:
        self._current_state = State.PROCESSING

    def sending(self) -> None:
        self._current_state = State.SENDING

    def reset(self) -> None:
        self._current_state      = State.INITIALIZATION
        self._command_to_execute = ""
